### PR TITLE
eth/tracers: add native flatCallTracer (aka parity style tracer)

### PR DESCRIPTION
**Pull request by ziogaschr**
This PR adds support for a flat call tracer, Similar to the parity tracing output, although with a few differences that are highlighted below.

**Implementation**: To minimize code duplication the `flatCallTracer` uses `callTracer` as a proxy to collect callstack info and at the end formats the output. Initial work on this PR had started here: https://github.com/etclabscore/ethereum.go-ethereum/pull/14

### Compatibility with Parity

1. Parity excludes calls to precompiles (when done via CALL and STATICCALL). We have maintained this behaviour, but added an option `includePrecompiles: true`, in which case all precompile calls are reported.
3. The error messages returned from parity and geth are different. If for compatibility reasons you want to receive parity-style errors pass the option `convertParityErrors: true`.
4. Parity removes the `result` object in case of failure. This behavior is maintained with the exception of reverts. Revert output usually contains useful information, i.e. Solidity revert reason.
5. In Parity `gasUsed` field of the main message does not account for intrinsic gas and refunds. Geth does.
6. When tracing a whole block Parity will return one result for whole block including one error. Geth will nest the tx results in a `result` field of the response object. And each tx will potentially have its own error.
7. Parity returns block and uncle rewards. Geth doesn't. Issuance is now handled by consensus clients, not EVM.

### CODE REVIEW

**Comment by s1na** [comment]
All tests passing in this branch: https://github.com/s1na/go-ethereum/commits/feat/flat-call-tracer. Some semantic differences between parity and our style we need to consider:

- Parity excludes calls to precompiles whereas we include. But weirdly they only exclude for CALL and STATICCALL. Chris suggested adding an option to let user decide. But we should stay consistent and either exclude or include for all CALL* variants.
- Delegatecall value is taken from its parent. in our call tracer we simply report zero for the value. We should consider adopting this change to our call tracer.
- The first frame of a parity trace reports gasUsed as we did before (not accounting for intrinsic gas and refunds at end of tx).

**Comment by s1na** [comment]
I refactored the result processing logic, i.e. taking the nested call frame, converting it to the flat format and encoding that:

- The encoding is now fully handled by gencodec. I.e. the flat structs store normal types and are overriden to hexutil types by gencodec. This allows us to add RLP encoding more easily in future.
- Many of the fields were pointers. I changed them back to value fields. I tried to only change in places where empty values don't need to be omitted from the JSON. But this is something we have to confirm. Which fields exactly need to be omitted when empty.

**Comment by ziogaschr** [comment]
@s1na sorry for missing out that the commits were on your fork.

- I really love you approach on using gencodec for handling the encoding.
- With regards struct fields being pointers, this was mostly because of reusing the same struct between tests and application code. In advance for some context fields, we had to use pointers as the context was available only in `debug_traceBlock` family calls and not in `debug_traceTransaction` family calls. To verify that everything is being set correctly and we remain compatible with the old Parity tracers, I will spin up an old OE instance and compare the full chain side-by-side with this PR.

**Comment by ziogaschr** [comment]
While running the comparison tests with **core-geth** tracers, I found the following incompatibilities so far (Goerli block reached = 5644).

> **core-geth** has been used for running those tests for the following reasons:
> - **core-geth** has been run the same test with **OE** on **ETC Mordor testnet** and it is compatible with **OE**
> - both **go-ethereum** and **core-geth** are able to sync with a common chain (**Goerli** is used), and in that sense we can use **core-geth** to run any tests for any on-chain transaction or even for new ones. **OE** isn't capable of syncing Goerli after merge, and even before that.

1. `debug_traceBlockByNumber` returns `txTraceResult` which adds a nested `result` field. 
  On `core-geth` the `trace_block` handles this for compatibility with OE. By keeping it as it is in **go-ethereum** the user is able to still get trace results with errors for each transaction in the block, while **OE** returns an error for the whole block.
2. Dealing with empty fields:
   - omit
     - `action.author`
     - `action.address`
     - `action.refundAddress`
     - `result.address`
   - keep
     - `action.gas`
     - `action.input`
     - `result.gasUsed`
     - `result.output`
   
  Those have been touched at https://github.com/ethereum/go-ethereum/pull/26377/commits/a402ba8b90259fb88d145e3595045fb1962e1dcd and I [said](https://github.com/ethereum/go-ethereum/pull/26377#issuecomment-1371111162) I will run tests and report any diffs.

3. `blockNumber` is a number and not hex in **OE**. We can keep it as hex for consistency with other methods.
4. `transactionPosition` is a number and not hex in **OE**.

---

@s1na I am going to handle 2 at the moment in order I can continue testing. For the rest, I will just tell the tests to ignore the changes, until we discuss.

**Comment by s1na** [comment]
Thanks for the fixes!

1. Yes I'd rather we keep the `traceBlock` API as is and not change it for a specific tracer.
2. I hate that now basically every field has to be a pointer. It can lead to weird behavior. For example I remember I had to copy the `from` and `to` addresses, but don't remember the exact bug I was trying to address (see https://github.com/ethereum/go-ethereum/pull/26377/commits/b2eeb0dc51d7d01032ddaa40a17715688d4868b6). I wish there was another way. Can you please copy the addrs as the mentioned commit?
3. Keep hex for consistency
4. Keep hex for consistency
5. Rewards have changed since the merge. There are no more uncles and block reward is handled by the beacon chain. Tho we could add a field for all gas fees going to the miner?

**Comment by ziogaschr** [comment]
2. I know what you mean, as I found this in the past too. I handle another case as well by this https://github.com/ethereum/go-ethereum/pull/26377/commits/af653978b4dafd123af14c7df2a426eb1e59f98b which covers what you mentioned. I still have to verify at Mordor, where we have this bug. Are you ok, handling it this way.

**Comment by s1na** [comment]
@ziogaschr I will think about an alternative approach. But I won't hold this PR because of it.

I updated the description of the PR to reflect all points on which we differ from Parity. Could you please add if I missed something? IMO 2. (delegated call value) and 5. (gas fees going to coinbase) can be considered. On other points I would stick with current behavior.

**Comment by ziogaschr** [comment]
Sounds great.
I will verify things in description and I will also verify that the code reflects them after I finish with the compatibility testing. In the same time I will note anything else found in code.

> 2. (delegated call value)

Can you please explain more on this one?

> 5. (gas fees going to coinbase)

I can see that this might be useful to have. Will you add this as a new result in the array?

**Comment by s1na** [comment]
> Can you please explain more on this one?

You cannot send value via DELEGATECALL. Similar to how the callee will inherit the storage of caller, it will also inherit the value that was sent to caller. Parity explicitly returns this value for delegatecall frames. But we historically returned 0 well because this call doesnt have a value of its own. But I can see the point in how parity does it.

> Will you add this as a new result in the array?

Maybe in the same field as block reward only with a slightly different meaning post-merge.

**Comment by ziogaschr** [comment]
@s1na I totally get it now for "2. (delegated call value)". I couldn't follow the context because of the "2." prefix, which felt like you were answering to point 2 from [the above list](https://github.com/ethereum/go-ethereum/pull/26377#issuecomment-1415197052)

File: eth/tracers/internal/tracetest/flat_calltrace_test.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+		Difficulty:  (*big.Int)(test.Context.Difficulty),
+		GasLimit:    uint64(test.Context.GasLimit),
+	}
+	_, statedb := tests.MakePreState(rawdb.NewMemoryDatabase(), test.Genesis.Alloc, false)
+
+	// txctx := &tracers.Context{
+	// 	BlockHash: blockHash,
+	// 	TxIndex:   task.index,
+	// 	TxHash:    txs[task.index].Hash(),
+	// }
```
Comment (s1na) [inline eth/tracers/internal/tracetest/flat_calltrace_test.go:110]:
```suggestion
```

File: eth/tracers/internal/tracetest/flat_calltrace_test.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	}
+	ret := new([]flatCallTrace)
+	if err := json.Unmarshal(res, ret); err != nil {
+		return fmt.Errorf("failed to unmarshal trace result: %v", err)
+	}
+	if !jsonEqualFlat(ret, test.Result) {
+		t.Logf("tracer name: %s", tracerName)
+
+		// uncomment this for easier debugging
+		have, _ := json.MarshalIndent(ret, "", " ")
```
Comment (s1na) [inline eth/tracers/internal/tracetest/flat_calltrace_test.go:142]: Ah I think I uncommented these for debugging and forgot to comment them again. Can you please comment before merging?

Comment (ziogaschr) [reply eth/tracers/internal/tracetest/flat_calltrace_test.go]: Resolved by https://github.com/ethereum/go-ethereum/pull/26377/commits/bb887f3e1c6b8e95758db14ef9b47572c4da6458

**Comment by s1na** [comment]
Team discussion: we should add delegate call value to the existing callTracer in a separate PR (as in our normal calltracer should also like parity return the parent's value for delegatecall frame).

However gas fees going to coinbase is unnecessary to add. It can be computed for each tx by gasPrice * gasUsed easily.

Otherwise PR is very good!

File: go.mod [hunk 3]
```diff
@@ -25,6 +25,7 @@ require (
 	github.com/gballet/go-libpcsclite v0.0.0-20190607065134-2772fd86a8ff
 	github.com/gballet/go-verkle v0.0.0-20220902153445-097bd83b7732
 	github.com/go-stack/stack v1.8.1
+	github.com/go-test/deep v1.0.8
```
[the remark this answers is not in the document]
Comment (ziogaschr) [reply go.mod:28]: Was being used here https://github.com/ethereum/go-ethereum/pull/26377/files#diff-d8ba7e48456ef054ab1561f6e13270b93b0d907ff1d5ad1abdb70e8057b78aaa @holiman. 

Will run a `go mod tidy`. Thanks

[the remark this answers is not in the document]
Comment (ziogaschr) [reply go.mod]: Resolved by [aebff89](https://github.com/ethereum/go-ethereum/pull/26377/commits/aebff894aa75e7d2c6113374e5d56611f28ea548)

**Comment by ziogaschr** [comment]
@s1na PR #26632 handles the delegatecall frame, though I removed this logic from this PR of flat call tracer. Based on this change a test fails, until #26632 is merged. After the merge, we can merge this PR to bring it in and have the tests pass. What else we can do, is to already merge #26632 in this one.

File: eth/tracers/native/call.go [hunk 4]
```diff
@@ -88,13 +88,30 @@ func (f *callFrame) processOutput(output []byte, err error) {
...
+}
+
 type callFrameMarshaling struct {
-	TypeString string `json:"type"`
-	Gas        hexutil.Uint64
-	GasUsed    hexutil.Uint64
-	Value      *hexutil.Big
-	Input      hexutil.Bytes
-	Output     hexutil.Bytes
+	Type    stringOpCode `json:"type"`
```
Comment (s1na) [inline eth/tracers/native/call.go:109]: I feel like you told me before what's the point of this change but I forgot. Could you please say again?

File: eth/tracers/native/call_flat.go [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+func (t *flatCallTracer) CaptureFault(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, depth int, err error) {
+	t.tracer.CaptureFault(pc, op, gas, cost, scope, depth, err)
+}
+
+// CaptureEnter is called when EVM enters a new scope (via call, create or selfdestruct).
+func (t *flatCallTracer) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	t.tracer.CaptureEnter(typ, from, to, input, gas, value)
+
+	// Child calls must have a value, even if it's zero.
+	if t.tracer.callstack[len(t.tracer.callstack)-1].Value == nil && value == nil {
```
Comment (s1na) [inline eth/tracers/native/call_flat.go:180]: After fixing this for delegate calls I think the only remaining case is STatic calls. I.e. static call passes in nil for the value to CaptureEnter. I'm wondering if we can pass in 0 instead?

File: eth/tracers/native/call_flat.go [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+type flatCallTracer struct {
+	tracer            *callTracer
+	config            flatCallTracerConfig
+	ctx               *tracers.Context // Holds tracer context data
+	reason            error            // Textual reason for the interruption
+	activePrecompiles []common.Address // Updated on CaptureStart based on given rules
+}
+
+type flatCallTracerConfig struct {
+	ConvertParityErrors bool `json:"convertParityErrors"` // If true, call tracer converts errors to parity format
```
[the remark this answers is not in the document]
Comment (ziogaschr) [reply eth/tracers/native/call_flat.go:124]: Sounds good. We kept it that way before in order to not break any automations anyone has made based on this errors. For example, some hardware wallet companies are utilising the tracers. For this reason, we thought that they might have a flow that matches on the actual string of an error.

I am going to remove the conversion.

File: eth/tracers/native/call_flat.go (hunk 5 above)
Comment (ziogaschr) [reply eth/tracers/native/call_flat.go:176]: Passing `0` is correct. On the other hand we don't have to match Parity that always sets a value for child calls, that said we can also remove this if and leave value to nil.

File: eth/tracers/native/call_flat.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ziogaschr) [reply eth/tracers/native/call_flat.go]: Resolved by [40df204](https://github.com/ethereum/go-ethereum/pull/26377/commits/40df20476327316243f643ac99bf91809581e232).

File: eth/tracers/native/call.go (hunk 4 above)
Comment (ziogaschr) [reply eth/tracers/native/call.go:109]: @s1na I am totally fine if you prefer reverting this change

File: eth/tracers/native/call_flat.go (hunk 5 above)
Comment (s1na) [reply eth/tracers/native/call_flat.go:146]: Let's keep this for now.

File: eth/tracers/native/call_flat.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (s1na) [reply eth/tracers/native/call_flat.go]: Thx!

File: eth/tracers/native/call_flat.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
+func (t *flatCallTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	t.tracer.CaptureStart(env, from, to, create, input, gas, value)
+}
+
+// CaptureEnd is called after the call finishes to finalize the tracing.
+func (t *flatCallTracer) CaptureEnd(output []byte, gasUsed uint64, err error) {
+	t.tracer.CaptureEnd(output, gasUsed, err)
+	// Parity trace considers only reports the gas used during the top call frame which doesn't include
```
Comment (s1na) [inline eth/tracers/native/call_flat.go:126]: Can you please revert this too? Our gas usage is more accurate, accounting for intrinsic gas and refunds too.

File: eth/tracers/native/call_flat.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+	Type                string          `json:"type"`
+}
+
+type flatCallFrameMarshaling struct {
+	BlockNumber         hexutil.Uint64
+	TransactionPosition hexutil.Uint64
+}
+
+type flatCallAction struct {
+	Author         *common.Address `json:"author,omitempty"`
```
Comment (s1na) [inline eth/tracers/native/call_flat.go:60]: These fields are not used in the code, we should remove them:

- Author
- RewardType

Also `CreationMethod` but we could fill that in (I suppose it should distinguish between txcreate, create, or create2).

**Pull request merged**